### PR TITLE
ENH: Fixed intercept computation in theilslopes

### LINKED ISSUE #14324: BUG: scipy.stats.theilslopes intercept calculation can produce incorrect results.
Theilslopes calculates the median intercept using the calculated theilslope estimate and the median x and median y values as: `np.median(y) - medslope * np.median(x)`. This can produce an incorrect intercept when there is a gap in the middle of the x values. Most other implementations of this algorithm use the median of all intercepts `np.median(y - medslope * x)`, which doesn't seem to have this issue.

The `np.median(y) - medslope * np.median(x)` is in the documentation, so I'm not entirely sure if there's a good reason for doing calculating the intercept that way.

#### Reproducing code example:
```
import pandas as pd
from scipy.stats import theilslopes
from io import StringIO
import matplotlib.pyplot as plt

test_data = StringIO("""
x,y
1,1
1.000507357,1.012345679
1.001014713,1.802469136
1.00152207,1.469135802
1.002029427,1.740740741
1.002536783,1.395061728
1.00304414,1.271604938
1.017250127,2.24691358
1.017757484,1.666666667
1.01826484,3.716049383
1.018772197,4.098765432
1.019279554,3.617283951
1.01978691,2.703703704
1.020294267,2.012345679
1.020801624,1.481481481
1.02130898,1.432098765
1.021816337,1.839506173
1.022323694,2.839506173
1.02283105,2.790123457
1.023338407,2.543209877
1.023845764,3.037037037
1.02435312,3.938271605
1.024860477,3.814814815
""")

df = pd.read_csv(test_data)

slopes = theilslopes(df.y, df.x)

df['line'] = df.x * slopes[0] + slopes[1]
df['line_corr'] = df.x * slopes[0] + (df.y - slopes[0] * df.x).median()

df.plot(x='x', y=['y', 'line', 'line_corr'], label=['Data', 'Theilslope intercept', 'Corrected Intercept'])
plt.show()
```
sample

#### Scipy/Numpy/Python version information:
```
import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info)
# 1.7.0 1.21.0 1.2.5 sys.version_info(major=3, minor=9, micro=5, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by czgdp1807**
#### Reference issue

Closes https://github.com/scipy/scipy/issues/14324

#### Additional information

ping @chrisb83

### CODE REVIEW

**Comment by czgdp1807** [comment]
@chrisb83 The tests are passing. Please review.

File: scipy/stats/_stats_mstats_common.py [hunk 1]
```diff
@@ -198,7 +198,7 @@ def linregress(x, y=None, alternative='two-sided'):
                             intercept_stderr=intercept_stderr)
 
 
-def theilslopes(y, x=None, alpha=0.95):
+def theilslopes(y, x=None, alpha=0.95, *, method='separate'):
```
Comment (chrisb83) [inline scipy/stats/_stats_mstats_common.py:201]: I think you can remove the *

File: scipy/stats/_stats_mstats_common.py [hunk 2]
```diff
@@ -215,13 +215,24 @@ def theilslopes(y, x=None, alpha=0.95):
...
         interpreted as "find the 90% confidence interval".
+    method : {'joint', 'separate'}, optional
+        Method to be used for computing estimate for intercept.
+        Following methods are supported,
+
+            * 'joint': Uses np.median(y - medslope * x) as intercept.
+            * 'separate': Uses np.median(y) - medslope * np.median(x)
+                          as intercept.
+
+        By default, 'separate'.
```
Comment (chrisb83) [inline scipy/stats/_stats_mstats_common.py:226]: The default is 'separate'.

File: scipy/stats/_stats_mstats_common.py [hunk 3]
```diff
@@ -215,13 +215,24 @@ def theilslopes(y, x=None, alpha=0.95):
...
+
+        .. versionadded:: 1.8.0
 
     Returns
     -------
     medslope : float
         Theil slope.
     medintercept : float
-        Intercept of the Theil line, as ``median(y) - medslope*median(x)``.
+        Intercept of the Theil line, as ``median(y - medslope*x)``.
```
Comment (chrisb83) [inline scipy/stats/_stats_mstats_common.py:235]: need to be reverted since we keep the default

File: scipy/stats/_stats_mstats_common.py [hunk 4]
```diff
@@ -234,8 +245,8 @@ def theilslopes(y, x=None, alpha=0.95):
     Notes
     -----
     The implementation of `theilslopes` follows [1]_. The intercept is
-    not defined in [1]_, and here it is defined as ``median(y) -
-    medslope*median(x)``, which is given in [3]_. Other definitions of
+    not defined in [1]_, and here it is defined as ``median(y -
```
Comment (chrisb83) [inline scipy/stats/_stats_mstats_common.py:248]: needs to be updated

File: scipy/stats/_stats_mstats_common.py [hunk 5]
```diff
@@ -298,7 +310,10 @@ def theilslopes(y, x=None, alpha=0.95):
     slopes = deltay[deltax > 0] / deltax[deltax > 0]
     slopes.sort()
     medslope = np.median(slopes)
-    medinter = np.median(y) - medslope * np.median(x)
+    if method == 'joint':
+        medinter = np.median(y - medslope * x)
+    else:
```
Comment (chrisb83) [inline scipy/stats/_stats_mstats_common.py:315]: write `elif method == 'separate': ... else raise ValueError('...')` or check that method is valid at the beginning of the function

File: scipy/stats/tests/test_stats.py [hunk 6]
```diff
@@ -1655,11 +1655,25 @@ def test_theilslopes():
     assert_almost_equal(slope, 0.5)
     assert_almost_equal(intercept, 0.5)
 
+    slope, intercept, lower, upper = stats.theilslopes([0, 1, 1],
```
Comment (chrisb83) [inline scipy/stats/tests/test_stats.py:1658]: one extra test case to be added that checks that an error is raised for invalid method like xyz => there should be some examples in test_stats.py

**Review by chrisb83 — COMMENTED** [review]
Thanks, this looks good except for a few minor details.

**Comment by czgdp1807** [comment]
@chrisb83 I have addressed all the reviews. Let me know if there is something else to do. Thanks.

**Comment by czgdp1807** [comment]
Seems like a CI time out.

**Comment by chrisb83** [comment]
> I have addressed all the reviews. Let me know if there is something else to do. Thanks.

Great, I pointed out a few more minor thing. once fixed, I will approve the PR.

**Comment by czgdp1807** [comment]
> I pointed out a few more minor thing. once fixed

I have addressed the new reviews.

**Comment by czgdp1807** [comment]
@chrisb83 Is there anything else to be done in this PR? Please let me know.

**Review by chrisb83 — COMMENTED** [review]
a few more minor points are open, almost there :)

**Comment by czgdp1807** [comment]
@chrisb83 All tests passing. Is it ready to go now?

**Review by chrisb83 — APPROVED** [review]
Thanks. All my feedback has been addressed and all tests pass.

**Pull request merged**